x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/formula/validation.ts

```diff
@@ -0,0 +1,688 @@
+  },
+  wrongFirstArgument: {
+    message: 'wrong first argument',
+    type: { operation: 'string', type: 'string', argument: 'any' as string | number },
+  },
+  cannotAcceptParameter: { message: 'cannot accept parameter', type: { operation: 'string' } },
+  shouldNotHaveField: { message: 'operation should not have field', type: { operation: 'string' } },
+  tooManyArguments: { message: 'too many arguments', type: { operation: 'string' } },
+  fieldWithNoOperation: {
+    message: 'unexpected field with no operation',
+    type: { field: 'string' },
+  },
+  failedParsing: { message: 'Failed to parse expression', type: { expression: 'string' } },
+  duplicateArgument: {
+    message: 'duplicate argument',
+    type: { operation: 'string', params: 'string' },
+  },
+  missingMathArgument: {
+    message: 'missing math argument',
+    type: { operation: 'string', count: 1, params: 'string' },
+  },
+};
+export const errorsLookup = new Set(Object.values(validationErrors).map(({ message }) => message));
+type ErrorTypes = keyof typeof validationErrors;
+type ErrorValues<K extends ErrorTypes> = typeof validationErrors[K]['type'];
+
+export interface ErrorWrapper {
+  message: string;
+  locations: TinymathLocation[];
+  severity?: 'error' | 'warning';
+}
+
+export function isParsingError(message: string) {
+  return message.includes(validationErrors.failedParsing.message);
+}
+
+function findFunctionNodes(root: TinymathAST | string): TinymathFunction[] {
+  function flattenFunctionNodes(node: TinymathAST | string): TinymathFunction[] {
+    if (!isObject(node) || node.type !== 'function') {
+      return [];
+    }
+    return [node, ...node.args.flatMap(flattenFunctionNodes)].filter(Boolean);
+  }
+  return flattenFunctionNodes(root);
+}
+
+export function hasInvalidOperations(
+  node: TinymathAST | string,
+  operations: Record<string, GenericOperationDefinition>
+): { names: string[]; locations: TinymathLocation[] } {
+  const nodes = findFunctionNodes(node).filter((v) => !isMathNode(v) && !operations[v.name]);
+  return {
+    // avoid duplicates
+    names: Array.from(new Set(nodes.map(({ name }) => name))),
+    locations: nodes.map(({ location }) => location),
+  };
+}
+
+export const getQueryValidationError = (
+  query: string,
+  language: 'kql' | 'lucene',
+  indexPattern: IndexPattern
+): string | undefined => {
+  try {
+    if (language === 'kql') {
+      esKuery.toElasticsearchQuery(esKuery.fromKueryExpression(query), indexPattern);
+    } else {
+      esQuery.luceneStringToDsl(query);
+    }
+    return;
+  } catch (e) {
+    return e.message;
+  }
+};
+
+function getMessageFromId<K extends ErrorTypes>({
+  messageId,
+  values: { ...values },
+  locations,
+}: {
+  messageId: K;
+  values: ErrorValues<K>;
+  locations: TinymathLocation[];
+}): ErrorWrapper {
+  let message: string;
+  // Use a less strict type instead of doing a typecast on each message type
+  const out = (values as unknown) as Record<string, string>;
```

Comment:
Not sure about this change.
The `ErrorValues<K>` should work together with the `switch/case`? Ideally the `switch` should not have `default` case based on that.

In case the typing is not required the `ErrorValues<K>` type can be removed and `values` can be set to `Record<string, string>`.

Reply:
@dej611 This change is required due to the way i18n is done in Kibana. The previous code that you wrote was invalid because it couldn't be statically analyzed by the checker.